### PR TITLE
Improvement for #13844

### LINKED ISSUE #13908: [FEATURE REQUEST] integer_log
In a recent PR #13844, the following function may have been useful.

```python
def integer_log(y, x):
    """Returns (e, bool) where e is the largest nonnegative integer
    such that y >= x**e and bool is True if y == x**e"""
    assert x > 1
    y = as_int(y)
    x = as_int(x)
    if x == 2:
        e = y.bit_length() - 1
        return e, 2**e == y
    r = e = 0
    while y >= x:
        d = x
        m = 1
        while y >= d:
            y, r = divmod(y, d)
            e += m
            if y > d:
                d *= d
                m *= 2
    return e, y == 1 and r == 0
```
As tested in the following, it is much faster than `int(log(y, x))`

```
def go(n, b):
    a = [randint(1, 2**123) for i in range(n)]
    b = [randint(2, b) for i in range(n)]
    v1 = []
    t = time()
    for i in range(n):
        v1.append(int(log(a[i], b[i])))
    intlog = time() - t
    v2 = []
    t = time()
    for i in range(n):
        v2.append(integer_log(a[i], b[i]))
    this = time()-t
    for i in range(n):
        try:
            assert v1[i] == v2[i][0]
        except:
            print(a[i], b[i])
            print(v1[i], v2[i])
            return
    return this/intlog
>>> go(100, 42)
0.0016...
```

This fills in a gap left by 
* `multiplicity` - what is the power of x in y?
* `perfect_power` - can y be written as x**e?
* `integer_nthroot` - what is the largest integer x <= y**(1/n)?
* `is_square` - is x a square number

### LINKED ISSUE #13922: solveset cannot solve multivariate expressions with symbols in exponents
```
>>> from sympy import *
>>> x, y = symbols('x y')
>>> solveset(2**x - y, x, domain = S.Reals)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 950, in solveset
    return _solveset(f, symbol, domain, _check=True)
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 739, in _solveset
    lhs, rhs_s = inverter(f, 0, symbol)
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 704, in <lambda>
    inverter = lambda f, rhs, symbol: inverter_func(f, rhs, symbol, domain)
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 122, in invert_real
    return _invert(f_x, y, x, domain)
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 104, in _invert
    x1, s = _invert_real(f_x, FiniteSet(y), x)
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 154, in _invert_real
    return _invert_real(h, imageset(Lambda(n, n - g), g_ys), symbol)
  File "/home/akash/sympy/sympy/solvers/solveset.py", line 196, in _invert_real
    if base is not S.Zero and rhs > 0:
  File "/home/akash/sympy/sympy/core/relational.py", line 229, in __nonzero__
    raise TypeError("cannot determine truth value of Relational")
TypeError: cannot determine truth value of Relational
```
`solveset` returned the following result prior to https://github.com/sympy/sympy/commit/811ffffc.
Example:
```
>>> solveset(2**x - y, x, domain = S.Reals)
Intersection(S.Reals, {log(y)/log(2)})
```

### PR DESCRIPTION
**Pull request by Yathartha22**
This PR intends to implement a much better way as solved in #13844.
`integer_log` is implemented and used instead of `integer_nthroot`  as discussed in #13908

fixes: #13908
closes: #13911
fixes: #13922
closes: #13925

ping @smichr

### CODE REVIEW

File: sympy/core/power.py [hunk 1]
```diff
@@ -99,6 +100,48 @@ def integer_nthroot(y, n):
...
+    ========
+    integer_nthroot
+    sympy.ntheory.primetest.is_square
+    sympy.ntheory.factor_.multiplicity
+    sympy.ntheory.factor_.perfect_power
+    """
+    if x <= 1:
+        raise AssertionError('x should have value greater than 1')
+    if y <= 0:
+        raise AssertionError('y should have value greater than 0')
```
Comment (ylemkimon) [inline sympy/core/power.py]: I think ValueError is more appropriate.

File: sympy/core/power.py [hunk 2]
```diff
@@ -99,6 +100,48 @@ def integer_nthroot(y, n):
...
+    if x == 2:
+        e = y.bit_length() - 1
+        return e, 2**e == y
+    r = e = 0
+    orig_y = y
+    while y >= x:
+        d = x
+        m = 1
+        while y >= d:
+            y, r = divmod(y, d)
```
Comment (ylemkimon) [inline sympy/core/power.py]: Why divmod is used?

Comment (smichr) [reply sympy/core/power.py]: See modified version on issue page.

File: sympy/solvers/tests/test_solveset.py [hunk 3]
```diff
@@ -730,6 +730,7 @@ def test_solveset_real_exp():
     assert solveset(Eq(-3**x, 27), x, S.Reals) == S.EmptySet
     assert solveset(Eq(-5**(x+1),27), x, S.Reals) == S.EmptySet
     assert solveset(Eq(2**(x-3), -16), x, S.Reals) == FiniteSet(7)
+    assert solveset(Eq(3**(x - 3), -3**39), x, S.Reals) == S.EmptySet
```
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py]: Isn't something like `Eq((-3)**(x - 3), -3**39)` more in line with the modifications?

File: sympy/core/power.py [hunk 4]
```diff
@@ -99,6 +100,60 @@ def integer_nthroot(y, n):
...
+        raise ValueError('y cannot take value as 0')
+
+    if x in (-2, 2):
+        x = int(x)
+        y = as_int(y)
+        e = y.bit_length() - 1
+        return e, x**e == y
+    if x < 0:
+        n, b = integer_log(y if y > 0 else -y, -x)
+        return n, b and bool(n % 2 if y < 0 else not n % 2)
```
Comment (Yathartha22) [inline sympy/core/power.py]: used `bool` for cases like: `integer_log(-49, -7)`

**Comment by Yathartha22** [comment]
Have a look @smichr . It turns out that there was slight confusion in interpreting the cases like:

```
>>> solveset(Eq(-2**x, 4), x, S.Reals)
EmptySet()
```
see [here](http://www.wolframalpha.com/input/?i=solve(-2**x+-+4))
and
```
>>> solveset(Eq((-2)**x, 4), x, S.Reals)
{2}
```
see [here](http://www.wolframalpha.com/input/?i=solve((-2)**x+-+4))

Consequently, the modifications fixes the  issue raised in #13925
ping @akash9712

**Comment by smichr** [comment]
+1

**Comment by Yathartha22** [comment]
Thanks @smichr

**Pull request merged**